File: Rust/ExampleTests/common/fandc_typed.rs
#![allow(dead_code, unused_imports, unused_variables)]

use super::json;
use super::fandc_string::FandCString;

#[derive(Debug, Clone, Default, PartialEq)]
pub struct FandCTyped {
    pub f: i32,
    pub c: i32,
    pub notes: String,
}

impl FandCTyped {
    pub fn from_str_struct(s: &FandCString) -> Self {
        Self {
            f: s.f.parse::<i32>().unwrap_or_default(),
            c: s.c.parse::<i32>().unwrap_or_default(),
            notes: s.notes.clone(),
        }
    }

    pub fn to_json_value(&self) -> json::Value {
        json::Value::Object(vec![
            ("f".to_string(), json::Value::number_from_i64(self.f as i64)),
            ("c".to_string(), json::Value::number_from_i64(self.c as i64)),
            ("notes".to_string(), json::Value::Str(self.notes.clone())),
        ])
    }

    pub fn to_json(&self) -> String {
        json::write(&self.to_json_value())
    }

    pub fn from_json_value(v: &json::Value) -> json::JsonResult<Self> {
        Ok(Self {
            f: json::as_i32(json::require(v, "f")?, "f")?,
            c: json::as_i32(json::require(v, "c")?, "c")?,
            notes: json::as_string(json::require(v, "notes")?, "notes")?,
        })
    }

    pub fn from_json(text: &str) -> json::JsonResult<Self> {
        Self::from_json_value(&json::parse(text)?)
    }

    pub fn to_json_list(list: &[FandCTyped]) -> String {
        json::write(&json::Value::Array(
            list.iter().map(|item| item.to_json_value()).collect(),
        ))
    }

    pub fn from_json_list(text: &str) -> json::JsonResult<Vec<FandCTyped>> {
        let root = json::parse(text)?;
        let items = json::as_array(&root, "FandCTyped")?;
        let mut result = Vec::with_capacity(items.len());
        for e in items {
            result.push(Self::from_json_value(e)?);
        }
        Ok(result)
    }
}

File: Swift/ExampleTests/common/FandCTyped.swift
public struct FandCTyped: Equatable, CustomStringConvertible {
    public let f: Int
    public let c: Int
    public let notes: String

    public init(f: Int, c: Int, notes: String) {
        self.f = f
        self.c = c
        self.notes = notes
    }

    public init(from s: FandCString) {
        self.f = Int(s.f) ?? 0
        self.c = Int(s.c) ?? 0
        self.notes = s.notes
    }

    public func toStringStruct() -> FandCString {
        return FandCString(
            f: String(describing: f),
            c: String(describing: c),
            notes: String(describing: notes)
        )
    }

    public static func toStringList(_ list: [FandCTyped]) -> [FandCString] {
        return list.map { $0.toStringStruct() }
    }

    public static func fromStringList(_ list: [FandCString]) -> [FandCTyped] {
        return list.map { FandCTyped(from: $0) }
    }

    public func toJSONValue() -> [String: Any] {
        return [
            "f": f,
            "c": c,
            "notes": notes,
        ]
    }

    public func toJSON() throws -> String {
        return try Json.write(toJSONValue())
    }

    public init(fromJSONValue m: [String: Any]) throws {
        self.f = try Json.asInt(Json.require(m, "f"), "f")
        self.c = try Json.asInt(Json.require(m, "c"), "c")
        self.notes = try Json.asString(Json.require(m, "notes"), "notes")
    }

    public init(fromJSON text: String) throws {
        try self.init(fromJSONValue: Json.parseObject(text))
    }

    public static func toJSONList(_ list: [FandCTyped]) throws -> String {
        return try Json.write(list.map { $0.toJSONValue() })
    }

    public static func fromJSONList(_ text: String) throws -> [FandCTyped] {
        return try Json.parseArray(text).map {
            try FandCTyped(fromJSONValue: Json.asObject($0, "FandCTyped"))
        }
    }

    public var description: String {
        return "F=\(f), C=\(c), Notes=\(notes)"
    }
}
